data/posts/2017-06-23-expanding-on-phalcon-3-2-0-new-features.md

```diff
@@ -0,0 +1,328 @@
+  
+  ```php
+  <?php
+
+  use Phalcon\Mvc\Model;
+
+  class User extends Model
+  {
+      public function initialize()
+      {
+          $this->keepSnapshots(true);
+      }
+  }
+  
+  $user = new User();
+  $user->name = "Test User";
+  $user->create();
+  var_dump($user->getChangedFields());
+  $user->login = "testuser";
+  var_dump($user->getChangedFields());
+  $user->update();
+  var_dump($user->getChangedFields());
+  ```
+  On Phalcon 3.0.0 output was:
+  ```
+  array(1) {
+    [0]=>
+    string(4) "name"
+  }
+  array(2) {
+    [0]=>
+    string(4) "name"
+    [1]=>
+    string(5) "login"
+  }
+  array(2) {
+    [0]=>
+    string(4) "name"
+    [1]=>
+    string(5) "login"
+  }
+  ```
+  
+  On Phalcon 3.1.0 and later it is:
+  ```
+  array(0) {
+  }
+  array(1) {
+    [0]=>
+    string(5) "login"
+  }
+  array(0) {
+  }
+  ```
+  
+  `Model::getUpdatedFields` will properly return updated fields or as mentioned above you can go back to previous behavior by changing ini value.
+  
+- ##### Added support for having option in `Phalcon\Paginator\Adapter\QueryBuilder` [#12111](https://github.com/phalcon/cphalcon/issues/12111)
+  From now you can use `Phalcon\Mvc\Model\Query\Builder::having` and `Phalcon\Paginator\Adapter\QueryBuilder`. Let's assume you have such a table `stock`:
+  
+  ```sql
+  DROP TABLE IF EXISTS `stock`;
+  CREATE TABLE `stock` (
+    `id` int(11) NOT NULL,
+    `name` varchar(32) NOT NULL,
+    `stock` int(11) NOT NULL
+  ) ENGINE=InnoDB DEFAULT CHARSET=latin1;
+  
+  LOCK TABLES `stock` WRITE;
+  INSERT INTO `stock` (`id`, `name`, `stock`) VALUES
+  (1, 'Apple', 2),
+  (2, 'Carrot', 6)
+  (3, 'pear', 0);
+  UNLOCK TABLES;
+  
+  ALTER TABLE `stock`
+    ADD PRIMARY KEY (`id`);
+  
+  ALTER TABLE `stock`
+    MODIFY `id` int(11) NOT NULL AUTO_INCREMENT, AUTO_INCREMENT=4;
+  ```
+  
+  And such a code:
+  
+  ```php
+  <?php
+
+  use Phalcon\Di\FactoryDefault;
+  use Phalcon\Paginator\Adapter\QueryBuilder;
+
+  class Stock extends Phalcon\Mvc\Model;
+
+  $di = new FactoryDefault();
+  $modelsManager = $di->get('modelsManager');
+  $builder = $modelsManager->createBuilder()
+      ->columns("*, COUNT(*) as stock_count")
+      ->from(['Stock' => Stock::class])
+      ->groupBy('name')
+      ->having('SUM(Stock.stock) > 0');
+
+  $paginate = (new QueryBuilder(
+      [
+          "builder" => $builder,
+          "limit"   => 1,
+          "page"    => 2
+      ]
+  ))->getPaginate();
+  var_dump($paginate->total_pages); // now it will return 2, previously it was 3
+  var_dump($paginate->total_items); // now it will return 2, previously it was 3
+  ```
+  
+  Keep in mind that MySQL need to select all rows and do count on it. Also we use `groupBy` to select columns for those rows. If you don't have `groupBy` then you need to pass `columns` option to `Phalcon\Paginator\Adapter\QueryBuilder`:
```

Comment:
Please note:
* MySQL will need to select all rows and do a `count` on it.
* We use `groupBy` to select columns for those rows.
* If you don't have `groupBy` then you need to pass `columns` option to `Phalcon\Paginator\Adapter\QueryBuilder`: